runner/sidecar/source/db/db.go

```diff
@@ -0,0 +1,217 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	dfv1 "github.com/argoproj-labs/argo-dataflow/api/v1alpha1"
+	"github.com/argoproj-labs/argo-dataflow/runner/sidecar/source"
+	sharedutil "github.com/argoproj-labs/argo-dataflow/shared/util"
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+var logger = sharedutil.NewLogger()
+
+var offsetTableSchema = `CREATE TABLE IF NOT EXISTS argo_dataflow_offsets (
+	table_name VARCHAR(255) NOT NULL,
+	consumer_group VARCHAR(255) NOT NULL,
+	offset VARCHAR(255),
+	PRIMARY KEY(table_name, consumer_group)
+)`
+
+type rowData = map[string]interface{}
+
+type dbSource struct {
+	db *sql.DB
+}
+
+func New(ctx context.Context, secretInterface corev1.SecretInterface, clusterName, namespace, pipelineName, stepName string, replica int, sourceName string, x dfv1.DBSource, f source.Func) (source.Interface, error) {
+	dataSource, err := getDataSource(ctx, secretInterface, x)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find data source: %w", err)
+	}
+	db, err := sql.Open(x.Driver, dataSource)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open database connection: %w", err)
+	}
+	db.SetConnMaxLifetime(time.Minute * 3)
+
+	if x.InitSchema {
+		_, err = db.ExecContext(ctx, offsetTableSchema)
+		if err != nil {
+			return nil, fmt.Errorf("failed to init offsets table schema: %w", err)
+		}
+	}
+
+	var offset string
+	consumerGroup := fmt.Sprintf("%s.%s.%s.%s.sources.%s", clusterName, namespace, pipelineName, stepName, sourceName)
+	offset, err = getOffsetFromDB(ctx, db, x.Table, consumerGroup)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			if _, err = insertOffset(ctx, db, x.Table, consumerGroup, ""); err != nil {
+				return nil, fmt.Errorf("failed to initialize offset: %w", err)
+			}
+		} else {
+			return nil, fmt.Errorf("failed to get offset from db: %w", err)
+		}
+	}
+
+	go func() {
+		for {
+			time.Sleep(x.PollInterval.Duration)
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				records, err := queryData(ctx, db, x.Table, x.OffsetColumn, offset)
+				if err != nil {
+					logger.Error(err, "failed to query data from table %q: %w", x.Table, err)
+					continue
+				}
+				for _, d := range records {
+					jsonData, err := json.Marshal(d)
+					if err != nil {
+						logger.Error(err, "failed to marshal to json: %w", err)
+						// skip
+						continue
+					}
+					if err := f(ctx, jsonData); err != nil {
+						logger.Error(err, "failed to process data: %w", err)
+						// skip
+						continue
+					}
+					offset = fmt.Sprintf("%v", d[x.OffsetColumn])
+				}
+			}
+		}
+	}()
+
+	// update offset in db
+	go func() {
+		ticker := time.NewTicker(x.CommitInterval.Duration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if offset != "" {
```

Comment:
Can we only update once? Eg by setting offset to empty string.

Reply:
it's only once.

Reply:
what if we have less than one item every 5s? how many update do we do?